pkg/provisioner/ironic/ironic.go

```diff
@@ -390,6 +383,165 @@ func (p *ironicProvisioner) InspectHardware() (result provisioner.Result, err er
+	var op nodes.UpdateOp
+	if imageSource, ok := ironicNode.InstanceInfo["image_source"]; !ok {
+		// no source, need to add
+		op = nodes.AddOp
+		p.log.Info("adding host settings in ironic")
+	} else if imageSource != p.host.Spec.Image.URL {
+		//  have a different source, need to update
+		op = nodes.ReplaceOp
+		p.log.Info("updating host settings in ironic")
+	}
+	if op != "" {
+		_, err = nodes.Update(
+			p.client,
+			ironicNode.UUID,
+			nodes.UpdateOpts{
+				nodes.UpdateOperation{
+					Op:    op,
+					Path:  "/instance_info/image_source",
+					Value: p.host.Spec.Image.URL,
+				},
+				nodes.UpdateOperation{
+					Op:    op,
+					Path:  "/instance_info/image_checksum",
+					Value: p.host.Spec.Image.Checksum,
+				},
+				// FIXME(dhellmann): We have to provide something for
+				// the disk size until
+				// https://storyboard.openstack.org/#!/story/2005165
+				// is fixed in ironic.
+				nodes.UpdateOperation{
+					Op:    op,
+					Path:  "/instance_info/root_gb",
+					Value: 10,
+				},
+				// FIXME(dhellmann): We need to specify the root
+				// device to receive the image. That should come from
+				// some combination of inspecting the host to see what
+				// is available and the hardware profile to give us
+				// instructions.
+				// nodes.UpdateOperation{
+				// 	Op:    nodes.AddOp,
+				// 	Path:  "/properties/root_device",
+				// 	Value: map[string]interface{},
+				// },
+			}).Extract()
+		if err != nil {
+			return result, errors.Wrap(err, "failed to update host settings in ironic")
+		}
+		p.status.State = statePreparingToProvision
+		result.Dirty = true
+		return result, nil
+	}
+
+	// Ironic has the settings it needs, see if it finds any issues
+	// with them.
+	if p.status.State == statePreparingToProvision {
+		ok, err := p.validateNode(ironicNode)
+		if err != nil {
+			return result, errors.Wrap(err, "could not validate host during registration")
+		}
+		if !ok {
+			p.status.State = stateValidationError
+			result.Dirty = true // validateNode() would have set the errors
+			return result, nil
+		}
+
+		// If validation is successful we can start moving the host
+		// through the states necessary to make it "available".
+		p.log.Info("making host available",
+			"lastError", ironicNode.LastError,
+			"current", ironicNode.ProvisionState,
+			"target", ironicNode.TargetProvisionState,
+		)
+		changeResult := nodes.ChangeProvisionState(
+			p.client,
+			ironicNode.UUID,
+			nodes.ProvisionStateOpts{
+				Target: nodes.TargetProvide,
+			})
+		if changeResult.Err != nil {
+			return result, errors.Wrap(changeResult.Err,
+				"failed to change provisioning state to provide")
+		}
+		p.status.State = stateMakingAvailable
+		result.Dirty = true
+		return result, nil
+	}
+
+	// Wait for the host to become available
+	if p.status.State == stateMakingAvailable {
+		if ironicNode.ProvisionState != nodes.Available {
+			p.log.Info("waiting for host to become available")
```

Comment:
What happens if a host never becomes available? For example, Ironic goes to `clean failed` - does this mean it'll just return dirty forever?

Reply:
Yeah, for this first pass I'm just trying to get 1 happy path working. There are going to be lots of error handling updates to make.